### PR TITLE
Upgrade: install intermediate pattern when needed

### PR DESCRIPTION
**Pull request by ereslibre**
## Why is this PR needed?

Do not install the intermediate pattern unless the upgrade of
Kubernetes is a major or minor one.

### CODE REVIEW

File: pkg/skuba/actions/node/upgrade/apply.go [hunk 1]
```diff
@@ -118,12 +118,14 @@ func Apply(target *deployments.Target) error {
 			if err != nil {
 				return err
 			}
-			err = target.Apply(deployments.KubernetesBaseOSConfiguration{
-				KubeadmVersion:    nodeVersionInfoUpdate.Update.APIServerVersion.String(),
-				KubernetesVersion: nodeVersionInfoUpdate.Current.APIServerVersion.String(),
-			}, "kubernetes.install-intermediate-node-pattern")
-			if err != nil {
-				return err
+			if nodeVersionInfoUpdate.HasMajorOrMinorUpdate() {
```
Comment (MaximilianMeister) [inline pkg/skuba/actions/node/upgrade/apply.go]: do we also have to put this around the `install-node-pattern` later after `kubeadm.upgrade.apply`? it's going to be introduced by https://github.com/SUSE/skuba/pull/585 

also possibly some other states don't have to be run if we don't have a major or minor update, e.g. restarting services?

Comment (ereslibre) [reply pkg/skuba/actions/node/upgrade/apply.go]: I prefer to play safe and improve later :)

**Comment by danielorf** [comment]
We seem to be running into LB pool quota:
```
1 error occurred:\n\t* openstack_lb_pool_v2.gangway_pool: Error creating pool: Expected HTTP response code [] when accessing [POST https://engcloud.prv.suse.net:9696/v2.0/lbaas/pools], but got 409 instead\n{"NeutronError": {"message": "Quota exceeded for resources: [\'pool\'].", "type": "OverQuota", "detail": ""}}\n
```

I am able to manually create loadbalancers in ECP though, so I'm not sure where this error is originating from.

File: internal/pkg/skuba/upgrade/node/versions.go [hunk 2]
```diff
@@ -35,6 +35,17 @@ type NodeVersionInfoUpdate struct {
 	Update  kubernetes.NodeVersionInfo
 }
 
+func (nviu NodeVersionInfoUpdate) HasMajorOrMinorUpdate() bool {
+	if nviu.Current.IsControlPlane() {
+		if nviu.Update.APIServerVersion.Major() > nviu.Current.APIServerVersion.Major() ||
+			nviu.Update.APIServerVersion.Minor() > nviu.Current.APIServerVersion.Minor() {
```
Comment (MaximilianMeister) [inline internal/pkg/skuba/upgrade/node/versions.go]: it's probably a very minor enhancement but as the major updates are very infrequent, we could check for minor first to speed up the check a little bit

Comment (ereslibre) [reply internal/pkg/skuba/upgrade/node/versions.go]: We can change that later, however that would mean negative logic on the pattern installation (`if !isPatchUpdate { installIntermediatePattern }`).

Comment (MaximilianMeister) [reply internal/pkg/skuba/upgrade/node/versions.go]: it shouldn't change any outside logic i meant like this:

```go
if nviu.Update.APIServerVersion.Minor() > nviu.Current.APIServerVersion.Minor() ||
        nviu.Update.APIServerVersion.Major() > nviu.Current.APIServerVersion.Major()
```

Comment (ereslibre) [reply internal/pkg/skuba/upgrade/node/versions.go]: Ah misunderstood. Did it in level of importance but yes I don't even know if Kubernetes 2.0 is planned :)

**Pull request merged**